### DISCUSSION TITLE
PAT Basic 1094. 谷歌的招聘 (20) (C语言实现) | PAT Solution in C

### DISCUSSION THREAD

**Discussion by xlucn** [xlucn/PAT#61] [General]
https://oliverlew.github.io/PAT/Basic/1094.html 

 题目

**Comment by sayakaichinose** [comment]
大佬，仿照您的方法做的，但是不知道为什么在prime函数里malloc的空间，在main函数中free后，有的测试点会出错，您可以看下吗 // Markdown 语法不太懂，抱歉...
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int isprime(int n);
char* prime(int L, int K, char* N);

int main (){
    int L, K;
    char N[1000];
    scanf("%d %d", &L, &K);
    scanf("%s", &N);
    char* ret = prime(L, K, N);
    printf("%s\n", ret);
//      free(ret);  会出错
//      ret = NULL;
    return 0;
}

char* prime(int L, int K, char* N){
    char* k = (char*) malloc(10*sizeof(char)); // 如何free掉?
    char* ret = "404";
    for(int i = 0; i <= L - K; ++i){
        strncpy(k, N+i, K);
        if(isprime(atoi(k))){
            ret = k;
            break;
        }
    }
    return ret;
}

int isprime(int num){
    int ret = 0;
    if(num == 2 || num == 3) ret = 1;
    if(num > 3){
        for(int i = 2; i * i <= num; ++i){
            if (num % i == 0){
                ret = 0;
                break;
            }else{
                ret = 1;
            }
        }   
    }
    return ret;
}
```

**Comment by xlucn** [comment]
> @sayakaichinose
> 大佬，仿照您的方法做的，但是不知道为什么在prime函数里malloc的空间，在main函数中free后，有的测试点会出错，您可以看下吗 // Markdown 语法不太懂，抱歉...

这样就够了，可读性比直接复制的强百倍 ：）

我不知道你得到的错误是哪一类（OJ会告诉你），也没太仔细分析，但这里有一个隐患：

> ```
> //      free(ret);  会出错
> ```

你free了`prime`函数的返回值，但在这个函数里，~你有可能返回的是一个常字符串`k`，也就是`404`~（错的，k不是404字符串）：

> ```
>         if(isprime(atoi(k))){
>             ret = k;
>             break;
>         }
> ```

然后，你malloc的内存就没释放。这至少是一个内存泄露，编程上是有bug的，但是会如何造成题目出错就不清楚了，你可以再自己看看。

**Comment by xlucn** [comment]
> @sayakaichinose

不对，我好像看反了，不好意思。现在暂时不清楚你的问题出在哪，但是这个指针换来换去的，感觉不太直观。。。

**Comment by sayakaichinose** [comment]
> @OliverLew
> > @sayakaichinose
> 
> 不对，我好像看反了，不好意思。现在暂时不清楚你的问题出在哪，但是这个指针换来换去的，感觉不太直观。。。

确实判断一下,再free就不出错了,  晚上找了半天错误都没看出来，感谢dalao :)
抱大腿
```
if(ret != "404"){
        free(ret);  
        ret = NULL;
    }
```
